### PR TITLE
(docs) Improve map function description.

### PR DESCRIPTION
**Pull request by gguillotte**
Clarify how the `map` function works and provide more context in the
examples.

### CODE REVIEW

**Comment by MikaelSmith** [comment]
Travis is enforcing a commit message of `(doc)` instead of `(docs)`.

**Comment by nfagerlund** [comment]
@MikaelSmith Why in the world would it do that? That's not the convention in this repo. (doc) has about 6 commits in the last year, and (docs) has more like 30-something.

**Comment by MikaelSmith** [comment]
Misunderstanding by @kylog and I probably.

**Comment by kylog** [comment]
@nfagerlund yep, misunderstanding. Would you prefer that we check for `(doc)` only or `(docs)` only or accept either? Ideally, it'd be nice to apply the same to (at least) puppet/facter/hiera.

**Comment by nfagerlund** [comment]
@kylog @MikaelSmith I think I'm cool with allowing both. I doubt people will want to do historical searches for all docs commits (such that we'd want consistency for easier searching), right? As far as I understand it, the use case for that commit tag is so you can tell when looking at some arbitrary commit whether it's a docs-only change, not for any kind of data mining. 

If we do want to enforce consistency, I'd vote for (docs) only.

**Comment by gguillotte** [comment]
Feedback should be incorporated.

File: lib/puppet/functions/map.rb [hunk 1]
```diff
@@ -1,33 +1,68 @@
...
-# @example Using map with two arguments
+# @example Using the `map` function with an array and a one-parameter lambda
 #
-#      # Turns hash into array of values
-#      $a.map |$key,$val|{ $val }
+# ~~~ puppet
+# # For the array $data, return an array containing the values
+# $data = [1,2,3]
+# $transformed_data = $data.map |$items| { $items }
+# # $transformed_data contains [1,2,3]
```
Comment (nfagerlund) [inline lib/puppet/functions/map.rb]: This example could be improved by actually transforming the data somehow. Like, `$data.map |$item| { $item * 10 }` (produces `[10, 20, 30]`).

File: lib/puppet/functions/match.rb [hunk 2]
```diff
@@ -1,19 +1,36 @@
-# Returns the match result of matching a String or Array[String] with one of:
+# Matches a regular expression against a string and returns an array containing the match
+# and any matched capturing groups.
 #
-# * Regexp
-# * String - transformed to a Regexp
-# * Pattern type
-# * Regexp type
+# The first argument a string or array of strings. The second argument is either a regular
```
Comment (nfagerlund) [inline lib/puppet/functions/match.rb]: Typo: should be "The first argument **is** a string or..."

**Comment by nfagerlund** [comment]
Two final comments; after that, :+1: :shipit: 

@kylog, would you prefer docs commits with revisions like this squashed into a single commit?

**Comment by kylog** [comment]
Yes, I'd recommend squashing these into a single commit, since it's most intelligibly presented to future readers of the commits as a single cohesive change. (If you wanted to separate the `map` changes from the `match` changes, you could squash to two commits along those lines. But I'd probably just squash to a single commit.)

Thanks for checking in, and these changes look really good!

**Comment by gguillotte** [comment]
Please let me know if I didn't squash these commits correctly.

**Comment by kylog** [comment]
@gguillotte no that's not squashed. Once you squash you'll have a single commit with a single commit message. Here's a how-to: http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html.  It will probably seem like an odd thing at first but once you get the hang of it, it's very handy.

You might grab someone to look over your shoulder when you do it the first few times. I'd offer but I'm working remote. But most of the devs around you have probably done it a bunch so ask around in hipchat or meatspace :)

**Comment by gguillotte** [comment]
Yep, that's the page I followed. I'll poke folks as I'm also remote today.

On Tue, Aug 11, 2015 at 8:17 AM, Kylo Ginsberg <notifications@github.com>
wrote:

> @gguillotte <https://github.com/gguillotte> no that's not squashed. Once
> you squash you'll have a single commit with a single commit message. Here's
> a how-to:
> http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html.
> It will probably seem like an odd thing at first but once you get the hang
> of it, it's very handy.
>
> You might grab someone to look over your shoulder when you do it the first
> few times. I'd offer but I'm working remote. But most of the devs around
> you have probably done it a bunch so ask around in hipchat or meatspace :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/puppetlabs/puppet/pull/4123#issuecomment-129926306>.
>



-- 
*PuppetConf 2015 <http://2015.puppetconf.com/> is coming to Portland,
Oregon! Join us October 5-9.*
*Register now to take advantage of the Early Bird discount
<https://www.eventbrite.com/e/puppetconf-2015-october-5-9-tickets-13115894995?discount=EarlyBird>
*
*—**save $249!*

**Comment by nfagerlund** [comment]
With this all squashed and ready, I'm :+1: on it.

**Pull request merged**